### PR TITLE
Optimize image copying in SW.safe_copy()

### PR DESCRIPTION
**Pull request by stacywsmith**
The SW.install() method uses the SW.safe_copy() method to copy
a software package to a Junos device if the no_copy argument is False.

This enhancement checks to see if the package is already present on the
Junos device and skips the copy if the package is present and the
file checksum of botht the local and remote packages match.

Also added additional arguments to control the copy and checksum
functionality.

### CODE REVIEW

**Comment by jnpr-community-netdev** [comment]
Can one of the admins verify this patch?

File: lib/jnpr/junos/utils/sw.py [hunk 1]
```diff
@@ -392,36 +417,44 @@ def _issu_nssu_requirement_validation(self):
...
+                raise ValueError('Unknown checksum algorithm: %s' %
+                                 (algorithm))
+            return rsp.findtext('.//checksum')
         except RpcError as e:
-
-            # e.errs is list of dictionaries
-            if hasattr(e, 'errs') and \
-                    list(filter(lambda x: 'No such file or directory' in
-                                x['message'], e.errs)):
+            if (hasattr(e, 'message') and
```
Comment (vnitinv) [inline lib/jnpr/junos/utils/sw.py]: in python 3.x, hasattr(e, 'message') might always give False

File: lib/jnpr/junos/utils/sw.py [hunk 2]
```diff
@@ -444,53 +480,89 @@ def safe_copy(self, package, **kvargs):
...
         if checksum is None:
-            _progress('computing local checksum on: %s' % package)
-            checksum = SW.local_md5(package)
+            _progress('computing checksum on local package: %s' % (package))
+            try:
+                checksum = SW.local_checksum(package,
+                                             algorithm=checksum_algorithm)
+            except IOError:
+                _progress('error computing checksum on local package: %s. '
+                          'Ensure the local package exists.' % (package))
```
Comment (vnitinv) [inline lib/jnpr/junos/utils/sw.py]: should we not return False from this exception?

Comment (stacywsmith) [inline lib/jnpr/junos/utils/sw.py]: Yes. Thanks. Fixed.

**Pull request merged**